File: backend/services/ItemService.js
/**
 * Item Service
 * Data access layer for itinerary item operations
 * Handles all database queries for items
 */

const db = require('../config/database');

class ItemService {
  /**
   * Find all items for a specific itinerary
   * @param {string} itineraryId - Itinerary UUID
   * @param {Object} options - Query options (category filter, sort)
   * @returns {Promise<Array>} - Array of items
   */
  static async findByItineraryId(itineraryId, options = {}) {
    const { category, sortBy = 'display_order', sortOrder = 'ASC' } = options;
    
    let query = `
      SELECT
        id,
        itinerary_id,
        category,
        title,
        description,
        start_datetime,
        end_datetime,
        location,
        confirmation_code,
        cost,
        currency,
        metadata,
        display_order,
        is_completed,
        created_at,
        updated_at
      FROM itinerary_items
      WHERE itinerary_id = $1
    `;
    
    const params = [itineraryId];
    
    if (category) {
      query += ' AND category = $2';
      params.push(category);
    }
    
    query += ` ORDER BY ${sortBy} ${sortOrder}`;
    
    const result = await db.query(query, params);
    return result.rows;
  }
  
  /**
   * Find single item by ID
   * @param {string} id - Item UUID
   * @returns {Promise<Object|null>} - Item object or null
   */
  static async findById(id) {
    const query = 'SELECT * FROM itinerary_items WHERE id = $1';
    const result = await db.query(query, [id]);
    return result.rows[0] || null;
  }
  
  /**
   * Create new item
   * @param {Object} data - Item data
   * @returns {Promise<Object>} - Created item
   */
  static async create(data) {
    const query = `
      INSERT INTO itinerary_items (
        itinerary_id,
        category,
        title,
        description,
        start_datetime,
        end_datetime,
        location,
        confirmation_code,
        cost,
        currency,
        metadata,
        display_order,
        is_completed
      )
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
      RETURNING *
    `;
    
    const values = [
      data.itinerary_id,
      data.category,
      data.title,
      data.description || null,
      data.start_datetime || null,
      data.end_datetime || null,
      data.location || null,
      data.confirmation_code || null,
      data.cost || null,
      data.currency || 'USD',
      data.metadata ? JSON.stringify(data.metadata) : null,
      data.display_order || 0,
      data.is_completed || false
    ];
    
    const result = await db.query(query, values);
    return result.rows[0];
  }
  
  /**
   * Update item
   * @param {string} id - Item UUID
   * @param {Object} data - Update data
   * @returns {Promise<Object|null>} - Updated item or null
   */
  static async update(id, data) {
    const fields = [];
    const values = [];
    let paramCount = 1;
    
    const updateableFields = [
      'category', 'title', 'description', 'start_datetime', 'end_datetime',
      'location', 'confirmation_code', 'cost', 'currency', 'metadata',
      'display_order', 'is_completed'
    ];
    
    updateableFields.forEach(field => {
      if (data[field] !== undefined) {
        fields.push(`${field} = $${paramCount++}`);
        values.push(field === 'metadata' && typeof data[field] === 'object'
          ? JSON.stringify(data[field])
          : data[field]);
      }
    });
    
    fields.push(`updated_at = NOW()`);
    
    if (fields.length === 1) {
      return this.findById(id);
    }
    
    values.push(id);
    
    const query = `
      UPDATE itinerary_items
      SET ${fields.join(', ')}
      WHERE id = $${paramCount}
      RETURNING *
    `;
    
    const result = await db.query(query, values);
    return result.rows[0] || null;
  }
  
  /**
   * Delete item
   * @param {string} id - Item UUID
   * @returns {Promise<boolean>} - True if deleted
   */
  static async delete(id) {
    const query = 'DELETE FROM itinerary_items WHERE id = $1';
    const result = await db.query(query, [id]);
    return result.rowCount > 0;
  }
  
  /**
   * Bulk delete items
   * @param {Array<string>} ids - Array of item UUIDs
   * @returns {Promise<number>} - Number of deleted items
   */
  static async bulkDelete(ids) {
    const query = 'DELETE FROM itinerary_items WHERE id = ANY($1)';
    const result = await db.query(query, [ids]);
    return result.rowCount;
  }
  
  /**
   * Reorder items
   * @param {Array<Object>} items - Array of {id, display_order}
   * @returns {Promise<boolean>} - True if successful
   */
  static async reorder(items) {
    return await db.transaction(async (client) => {
      for (const item of items) {
        await client.query(
          'UPDATE itinerary_items SET display_order = $1, updated_at = NOW() WHERE id = $2',
          [item.display_order, item.id]
        );
      }
      return true;
    });
  }
  
  /**
   * Get itinerary ID for an item
   * @param {string} itemId - Item UUID
   * @returns {Promise<string|null>} - Itinerary UUID or null
   */
  static async getItineraryId(itemId) {
    const query = 'SELECT itinerary_id FROM itinerary_items WHERE id = $1';
    const result = await db.query(query, [itemId]);
    return result.rows[0]?.itinerary_id || null;
  }
  
  /**
   * Count items by category for an itinerary
   * @param {string} itineraryId - Itinerary UUID
   * @returns {Promise<Object>} - Category counts
   */
  static async countByCategory(itineraryId) {
    const query = `
      SELECT
        category,
        COUNT(*) as count
      FROM itinerary_items
      WHERE itinerary_id = $1
      GROUP BY category
    `;
    
    const result = await db.query(query, [itineraryId]);
    
    const counts = {
      accommodation: 0,
      activity: 0,
      transport: 0
    };
    
    result.rows.forEach(row => {
      counts[row.category] = parseInt(row.count);
    });
    
    return counts;
  }
  
  /**
   * Toggle item completion status
   * @param {string} id - Item UUID
   * @returns {Promise<Object|null>} - Updated item or null
   */
  static async toggleCompletion(id) {
    const query = `
      UPDATE itinerary_items
      SET is_completed = NOT is_completed, updated_at = NOW()
      WHERE id = $1
      RETURNING *
    `;
    
    const result = await db.query(query, [id]);
    return result.rows[0] || null;
  }
}

module.exports = ItemService;

File: backend-dotnet/VacationPlan.Core/Services/ItemService.cs
using VacationPlan.Core.DTOs;
using VacationPlan.Core.Interfaces;
using VacationPlan.Core.Models;

namespace VacationPlan.Core.Services;

/// <summary>
/// Service implementation for Itinerary Item business logic
/// Handles business rules, validation, and DTO mapping
/// </summary>
public class ItemService : IItemService
{
    private readonly IItemRepository _itemRepository;
    private readonly IItineraryRepository _itineraryRepository;

    public ItemService(
        IItemRepository itemRepository,
        IItineraryRepository itineraryRepository)
    {
        _itemRepository = itemRepository;
        _itineraryRepository = itineraryRepository;
    }

    /// <summary>
    /// Get all items for an itinerary with optional category filtering
    /// </summary>
    public async Task<ListResponse<ItemResponseDto>> GetItemsByItineraryAsync(
        Guid itineraryId,
        Guid userId,
        string? category = null)
    {
        // Verify user owns the itinerary
        var exists = await _itineraryRepository.ExistsAsync(itineraryId, userId);
        if (!exists)
            throw new UnauthorizedAccessException("Itinerary not found or access denied");

        var items = await _itemRepository.GetItemsByItineraryAsync(itineraryId, category);

        var response = items.Select(MapToDto).ToList();

        return new ListResponse<ItemResponseDto>
        {
            Data = response,
            Count = response.Count
        };
    }

    /// <summary>
    /// Get a single item by ID
    /// </summary>
    public async Task<ItemResponseDto?> GetItemByIdAsync(Guid id, Guid userId)
    {
        var item = await _itemRepository.GetByIdAsync(id);
        
        if (item == null)
            return null;

        // Verify user owns the parent itinerary
        var hasAccess = await _itineraryRepository.ExistsAsync(item.ItineraryId, userId);
        if (!hasAccess)
            throw new UnauthorizedAccessException("Access denied");

        return MapToDto(item);
    }

    /// <summary>
    /// Create a new item
    /// </summary>
    public async Task<ItemResponseDto> CreateItemAsync(
        Guid itineraryId,
        Guid userId,
        CreateItemDto dto)
    {
        // Verify user owns the itinerary
        var exists = await _itineraryRepository.ExistsAsync(itineraryId, userId);
        if (!exists)
            throw new UnauthorizedAccessException("Itinerary not found or access denied");

        // Business rule: Validate category
        var validCategories = new[] { "accommodation", "activity", "transport" };
        var normalizedCategory = dto.Category.ToLower();
        if (!validCategories.Contains(normalizedCategory))
            throw new ArgumentException($"Invalid category. Must be one of: {string.Join(", ", validCategories)}");

        // Business rule: Validate date range if provided
        if (dto.StartDatetime.HasValue && dto.EndDatetime.HasValue)
        {
            if (dto.EndDatetime.Value < dto.StartDatetime.Value)
                throw new ArgumentException("End datetime must be after start datetime");
        }

        var item = new ItineraryItem
        {
            ItineraryId = itineraryId,
            Category = normalizedCategory,
            Title = dto.Title,
            Description = dto.Description,
            StartDatetime = dto.StartDatetime,
            EndDatetime = dto.EndDatetime,
            Location = dto.Location,
            ConfirmationCode = dto.ConfirmationCode,
            Cost = dto.Cost,
            Currency = dto.Currency ?? "USD",
            MetadataDict = dto.Metadata,
            DisplayOrder = dto.DisplayOrder ?? 0
        };

        var created = await _itemRepository.CreateAsync(item);
        return MapToDto(created);
    }

    /// <summary>
    /// Update an existing item
    /// </summary>
    public async Task<ItemResponseDto?> UpdateItemAsync(
        Guid id,
        Guid userId,
        UpdateItemDto dto)
    {
        var item = await _itemRepository.GetByIdAsync(id);

        if (item == null)
            return null;

        // Verify user owns the parent itinerary
        var hasAccess = await _itineraryRepository.ExistsAsync(item.ItineraryId, userId);
        if (!hasAccess)
            throw new UnauthorizedAccessException("Access denied");

        // Apply updates
        if (!string.IsNullOrEmpty(dto.Title))
            item.Title = dto.Title;
    /// <summary>
    /// Bulk operations (reorder, bulk_delete)
    /// </summary>
    public async Task<BulkOperationResponse> BulkOperationAsync(Guid userId, BulkOperationDto dto)
    {
        if (dto.Operation.ToLower() == "reorder")
        {
            if (dto.Items == null || dto.Items.Count == 0)
                throw new ArgumentException("Items list is required for reorder operation");

            // Verify all items belong to user
            foreach (var item in dto.Items)
            {
                var existingItem = await _itemRepository.GetByIdAsync(item.Id);
                if (existingItem == null)
                    throw new ArgumentException($"Item {item.Id} not found");

                var hasAccess = await _itineraryRepository.ExistsAsync(existingItem.ItineraryId, userId);
                if (!hasAccess)
                    throw new UnauthorizedAccessException($"Access denied for item {item.Id}");
            }

            var reorderList = dto.Items.Select(i => (i.Id, i.DisplayOrder)).ToList();
            await _itemRepository.BulkReorderAsync(reorderList);

            return new BulkOperationResponse
            {
                Message = "Bulk reorder completed successfully",
                Processed = dto.Items.Count
            };
        }
        else if (dto.Operation.ToLower() == "bulk_delete")
        {
            if (dto.ItemIds == null || dto.ItemIds.Count == 0)
                throw new ArgumentException("ItemIds list is required for bulk_delete operation");

            // Verify all items belong to user
            foreach (var itemId in dto.ItemIds)
            {
                var item = await _itemRepository.GetByIdAsync(itemId);
                if (item == null)
                    throw new ArgumentException($"Item {itemId} not found");

                var hasAccess = await _itineraryRepository.ExistsAsync(item.ItineraryId, userId);
                if (!hasAccess)
                    throw new UnauthorizedAccessException($"Access denied for item {itemId}");
            }

            await _itemRepository.BulkDeleteAsync(dto.ItemIds);

            return new BulkOperationResponse
            {
                Message = "Bulk delete completed successfully",
                Processed = dto.ItemIds.Count
            };
        }
        else
        {
            throw new ArgumentException($"Invalid operation: {dto.Operation}. Supported operations: reorder, bulk_delete");
        }
    }

    {
        var item = await _itemRepository.GetByIdAsync(id);

        if (item == null)
            return false;

        // Verify user owns the parent itinerary
        var hasAccess = await _itineraryRepository.ExistsAsync(item.ItineraryId, userId);
        if (!hasAccess)
            throw new UnauthorizedAccessException("Access denied");

        await _itemRepository.DeleteAsync(id);
        return true;
    }

    /// <summary>
    /// Bulk update items (reorder, delete multiple)
    /// </summary>
    public async Task<bool> BulkUpdateItemsAsync(Guid userId, BulkUpdateItemsDto dto)
    {
        // Verify user owns all items
        foreach (var operation in dto.Operations)
        {
            var item = await _itemRepository.GetByIdAsync(operation.ItemId);
            if (item == null)
                throw new ArgumentException($"Item {operation.ItemId} not found");

            var hasAccess = await _itineraryRepository.ExistsAsync(item.ItineraryId, userId);
            if (!hasAccess)
                throw new UnauthorizedAccessException($"Access denied for item {operation.ItemId}");
        }

        // Process operations
        foreach (var operation in dto.Operations)
        {
            switch (operation.Action.ToLower())
            {
                case "delete":
                    await _itemRepository.DeleteAsync(operation.ItemId);
                    break;
                    
                case "reorder":
                    if (operation.DisplayOrder.HasValue)
                    {
                        var item = await _itemRepository.GetByIdAsync(operation.ItemId);
                        if (item != null)
                        {
                            item.DisplayOrder = operation.DisplayOrder.Value;
                            await _itemRepository.UpdateAsync(item);
                        }
                    }
                    break;
                    
                default:
                    throw new ArgumentException($"Invalid action: {operation.Action}");
            }
        }

        return true;
    }

    /// <summary>
    /// Map ItineraryItem entity to DTO
    /// </summary>
    private static ItemResponseDto MapToDto(ItineraryItem item)
    {
        return new ItemResponseDto
        {
            Id = item.Id,
            ItineraryId = item.ItineraryId,
            Category = item.Category,
            Title = item.Title,
            Description = item.Description,
            StartDatetime = item.StartDatetime,
            EndDatetime = item.EndDatetime,
            Location = item.Location,
            ConfirmationCode = item.ConfirmationCode,
            Cost = item.Cost,
            Currency = item.Currency,
            Metadata = item.MetadataDict,
            DisplayOrder = item.DisplayOrder,
            IsCompleted = item.IsCompleted,
            CreatedAt = item.CreatedAt,
            UpdatedAt = item.UpdatedAt
        };
    }
}
